### PR TITLE
[DAG] SimplifyDemandedVectorElts - add handling for INT<->FP conversions

### PR DESCRIPTION
**Pull request by RKSimon**

### CODE REVIEW

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll [hunk 1]
```diff
@@ -141,56 +141,61 @@ declare <8 x half> @llvm.ldexp.v8f16.v8i16(<8 x half>, <8 x i16>)
 define <8 x half> @fmul_pow2_8xhalf(<8 x i16> %i) {
 ; CHECK-SSE-LABEL: fmul_pow2_8xhalf:
 ; CHECK-SSE:       # %bb.0:
-; CHECK-SSE-NEXT:    subq $88, %rsp
-; CHECK-SSE-NEXT:    .cfi_def_cfa_offset 96
+; CHECK-SSE-NEXT:    subq $104, %rsp
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:144]: Minor regression due to complexity the scalarisation of the f16 conversions - I'm not convinced this is a showstopper tbh.

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll [hunk 2]
```diff
@@ -141,56 +141,61 @@ declare <8 x half> @llvm.ldexp.v8f16.v8i16(<8 x half>, <8 x i16>)
...
 ; CHECK-SSE-NEXT:    movdqa %xmm0, {{[-0-9]+}}(%r{{[sb]}}p) # 16-byte Spill
-; CHECK-SSE-NEXT:    pshufd {{.*#+}} xmm0 = xmm0[1,1,1,1]
+; CHECK-SSE-NEXT:    psrld $16, %xmm0
 ; CHECK-SSE-NEXT:    cvtdq2ps %xmm0, %xmm0
 ; CHECK-SSE-NEXT:    callq __truncsfhf2@PLT
 ; CHECK-SSE-NEXT:    movss %xmm0, {{[-0-9]+}}(%r{{[sb]}}p) # 4-byte Spill
-; CHECK-SSE-NEXT:    cvtdq2ps {{[-0-9]+}}(%r{{[sb]}}p), %xmm0 # 16-byte Folded Reload
+; CHECK-SSE-NEXT:    movdqa {{[-0-9]+}}(%r{{[sb]}}p), %xmm0 # 16-byte Reload
+; CHECK-SSE-NEXT:    psrlq $48, %xmm0
+; CHECK-SSE-NEXT:    cvtdq2ps %xmm0, %xmm0
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:168]: The bigger problem appears to be all these cvtdq2ps(shuffle) calls that could be replaced with shuffle(cvtdq2ps) and share the same cvtdq2ps

File: llvm/test/CodeGen/PowerPC/pr38087.ll [hunk 3]
```diff
@@ -11,9 +11,9 @@ declare { i32, i1 } @llvm.usub.with.overflow.i32(i32, i32) #0
 define void @draw_llvm_vs_variant0(<4 x float> %x) {
 ; CHECK-LABEL: draw_llvm_vs_variant0:
 ; CHECK:       # %bb.0: # %entry
-; CHECK-NEXT:    lxsd v3, 0(r3)
-; CHECK-NEXT:    vmrghh v3, v3, v3
+; CHECK-NEXT:    lxsihzx v3, 0, r3
 ; CHECK-NEXT:    vextsh2w v3, v3
+; CHECK-NEXT:    xxmrghw v3, v3, v3
```
Comment (RKSimon) [inline llvm/test/CodeGen/PowerPC/pr38087.ll:16]: @stefanp-ibm I'm assuming this change is neutral - is that correct?

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 4]
```diff
@@ -3736,6 +3736,14 @@ bool TargetLowering::SimplifyDemandedVectorElts(
...
     }
     break;
+  case ISD::SINT_TO_FP:
+  case ISD::UINT_TO_FP:
+  case ISD::FP_TO_SINT:
+  case ISD::FP_TO_UINT:
+    if (SimplifyDemandedVectorElts(Op.getOperand(0), DemandedElts, KnownUndef,
+                                   KnownZero, TLO, Depth + 1))
+      return true;
+    break;
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:3746]: Is the fallthrough to the undef handling tested (not sure this is actually a useful path)

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:3746]: Nice catch - int2fp should fold undef to zero - I'll just early out for now

**Review by arsenm — APPROVED** [review]

**Pull request merged**